### ISSUE TITLE
CommunicationError : Error while copying content to a stream

### ISSUE THREAD

**Issue by spookymattress** [microsoft/navcontainerhelper#463]
I have an issue with the **Run-TestsInNavContainer** method. It throws the following error:
```
CommunicationError : Error while copying content to a stream.
ClientSession in Error
at AwaitState, C:\ProgramData\NavContainerHelper\Extensions\BC14DK\PsTestTool\ClientContext.ps1: line 98
at InvokeInteraction, C:\ProgramData\NavContainerHelper\Extensions\BC14DK\PsTestTool\ClientContext.ps1: line 111
at InvokeInteractionAndCatchForm, C:\ProgramData\NavContainerHelper\Extensions\BC14DK\PsTestTool\ClientContext.ps1: line 120
at OpenForm, C:\ProgramData\NavContainerHelper\Extensions\BC14DK\PsTestTool\ClientContext.ps1: line 135
at Run-Tests, C:\ProgramData\NavContainerHelper\Extensions\BC14DK\PsTestTool\PsTestFunctions.ps1: line 132
at <ScriptBlock>, <No file>: line 38
ClientSession in Error
At C:\Program Files\WindowsPowerShell\Modules\navcontainerhelper\0.6.1.0\ContainerHandling\Invoke-ScriptInNavContainer.ps1:31 char:13
+             Invoke-Command -Session $session -ScriptBlock $scriptbloc ...
+             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (ClientSession in Error:String) [], RuntimeException
    + FullyQualifiedErrorId : ClientSession in Error
```

I think it is related to the fact that I am running Windows 10 Version 1903, so the host os is not the same as the container os? Therefore I am also using hyper-v

Container OS Version: 10.0.17763.437 (ltsc2019)
Host OS Version: 10.0.18362.116 (1903)

Could this be the issue?

In your blog post you write:

> So, as soon as 1903 is publicly available we will have a generic image for 1903 and new-navcontainer will be able to re-platform ltsc2019 containers to 1903 on-the-fly and start them with process isolation if you use -useBestContainerOS.

Any idea when these will be available?

**Comment by freddydk** [comment]
Generic image is available for 1903 now and NavContainerHelper 0.6.1.2 supports this with -useBestContainerOS (i.e. process isolation for 1903 is available through this)

**Comment by kine** [comment]
Having same issues, sometimes it is ok, some times I get this error. Just re-starting the pipeline will mostly help. It means it is instability, which raise in my case in ration 1:4 (failing:ok).

In my case it is

**Comment by freddydk** [comment]
Will have a look at this asap

**Comment by kine** [comment]
It is not question of last versions, it was there for some time, but right now I am hitting it more because generating more builds (testing multistage yaml pipelines with release stages).

**Comment by kine** [comment]
Same for

Container OS Version: 10.0.17763.503 (ltsc2019)
Host OS Version: 10.0.17763.107 (ltsc2019)

**Comment by freddydk** [comment]
I am trying to repro this. I am running this script:

```
$auth = "NavUserPassword"
$credential = New-Object pscredential 'admin', (ConvertTo-SecureString -String 'P@ssword1' -AsPlainText -Force)

New-BcContainer -accept_eula `
                -accept_outdated `
                -containerName test `
                -imageName mcr.microsoft.com/businesscentral/onprem:1904-cu1-w1 `
                -auth $auth `
                -Credential $credential `
                -updateHosts `
                -licenseFile $mylicenseFile `
                -includeTestToolkit `
                -includeCSide `
                -doNotExportObjectsToText

Publish-NavContainerApp -containerName test -appFile "https://www.dropbox.com/s/uiuldvj73zxkofj/Default%20publisher_runalltests_1.0.0.0.app?dl=1" -skipVerification -sync -install

Run-TestsInNavContainer -containerName test -detailed -credential $credential -XUnitResultFileName 'c:\programdata\navcontainerhelper\results.xml'
```
and... - I have failing tests, but I haven't seen the ClientSessionInError.
What would be really nice is, to get a script like that, which would fail (just sometimes)

**Comment by freddydk** [comment]
I managed to get a repro and I found out what goes wrong.
Currently there is a hardcoded transaction timeout of 10 minutes.
This means that if one test codeunit takes more than 10 minutes, the session will go into error state and the remaining tests won't run.
That also means that a test with a neverending loop will in-effect block the remaining test runs.
It also means that slow machines are more prone to have bad tests.
Putting this together with the fact, that I wan't to implement test re-runs - I decided to fix these two issues at the same time.

Next NavContainerHelper, you will be able to do:

```
$tests = Get-TestsFromBCContainer -containerName test -credential $credential

$XUnitResultFileName = 'c:\programdata\navcontainerhelper\results.xml'
Remove-Item $XUnitResultFileName -Force -ErrorAction Ignore

$runTestsParameters = @{
    "containerName" = "test"
    "detailed" = $true
    "credential" = $credential
    "XUnitResultFileName" = $XUnitResultFileName
    "AppendToXUnitResultFile" = $true
}

$tests.Codeunits | % {

    $allPassed = Run-TestsInNavContainer @runTestsParameters -testCodeunit $_.Id
    if (-not $allPassed) {
        Run-TestsInNavContainer @runTestsParameters -ReRun -testCodeunit $_.Id
    }
}
```

You can also collect failing codeunits and recreate the container and run the failing codeunit on that.
This means that every codeunit will be a new session and if one codeunit times out - it won't block anything.
You could also investigate the XUnit xml file to see which tests have failed or hasn't run yet.
Comments on this idea?

**Comment by freddydk** [comment]
Should be fixed in 0.6.2.1 - just released.
